tabgroups/init.lua

```diff
@@ -0,0 +1,870 @@
+--- Tab groups management module
+--
+-- This module allows you to group opened tabs and switch between different groups
+-- and tabs in groups
+--
+-- # Capabilities
+--
+-- # Usage
+--
+-- * Add `require "tabgroups"` to your `config.rc`.
+-- * Press 'x' to open list of defined tabgroups
+-- * Press 'X' to open list of tabs in active tabgroup
+--
+-- # Troubleshooting
+--
+-- # Files and Directories
+--
+--
+-- @module tabgroups
+-- @author Serg Kozhemyakin <serg.kozhemyakin@gmail.com>
+-- @author Aidan Holm <aidanholm@gmail.com>
+-- @copyright 2017 Serg Kozhemyakin <serg.kozhemyakin@gmail.com>
+
+local string = require("string")
+local window = require("window")
+local webview = require("webview")
+local binds, modes = require("binds"), require("modes")
+local add_binds = modes.add_binds
+local menu_binds = binds.menu_binds
+local new_mode = require("modes").new_mode
+local session = require("session")
+local settings = require("settings")
+local lousy = require("lousy")
+
+local _new_tabgroup_prefix = "Unnamed#"
+local _default_notify = "n: create new group, d: delete group, r: rename group"
+
+-- private hash for storing map uri to tabgroup
+local w2groups = setmetatable({}, { __mode = "k" })
+
+-- temporary table for storing deleted notebooks, will be cleaned at idle time
+local _deleted_groups = {}
+
+local switch_tabgroup, delete_tabgroup
+
+local function _get_next_tabgroup_name(w)
+    local name
+    local i = 1
+    if w2groups[w] and w2groups[w].groups then
+        repeat
+            i = i + 1
+            name = _new_tabgroup_prefix .. i
+        until not w2groups[w].groups[name]
+    else
+        name = _new_tabgroup_prefix .. i
+    end
+    return name
+end
+
+local function grouptabs(w, group)
+    assert(type(group) == 'string')
+    assert(w2groups[w].groups[group])
+    assert(w2groups[w].groups[group]._notebook)
+    assert(w2groups[w].groups[group]._notebook.type == "notebook")
+    local i = 0
+    local n = w2groups[w].groups[group]._notebook:count()
+    return function()
+        i = i + 1
+        if i <= n then return i, w2groups[w].groups[group]._notebook[i] end
+    end
+end
+
+local function webview2idx(w, group, view)
```

Comment:
Can't this just be `view.parent:indexof(view)`? With separate handling if `view.parent ~= w2groups[w].groups[group]._notebook` of course